### PR TITLE
Add thread option to samtools modules

### LINKED ISSUE #1066: Allow argument options and set the number of threads for samtools merge and other subtools
## Is your feature request related to a problem? Please describe
Current samtools/merge doesn't allow argument options and set the number of threads

## Describe the solution you'd like
Change the command to allow argument options : 
From 

```bash
samtools merge  ${reference} ${prefix}.${file_type} $input_files
```
To
```bash
samtools merge $options.args -@ $task.cpus  ${reference} ${prefix}.${file_type} $input_files
```

### LINKED ISSUE #1070: [BUG] Update samtools version to 1.14
## Description of the bug

Samtools had a new release to update to

### PR DESCRIPTION
**Pull request by FriederikeHanssen**
Closes #1066 
Closes #1070 

## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

**Comment by phue** [comment]
Hi,
I'm not entirely sure if that causes real-world issues but according to the samtools docs:
```
-@, --threads INT

    Number of input/output compression threads to use in addition to main thread
```

From my understanding, using `-@ task.cpus` would then actually try to use `task.cpus + 1` cores

**Comment by FriederikeHanssen** [comment]
> Hi, I'm not entirely sure if that causes real-world issues but according to the samtools docs:
> 
> ```
> -@, --threads INT
> 
>     Number of input/output compression threads to use in addition to main thread
> ```
> 
> From my understanding, using `-@ task.cpus` samtools would then actually try to use `task.cpus + 1` cores

Ah interesting. I will try to figure out if that is actually the case

File: modules/samtools/view/main.nf [hunk 1]
```diff
@@ -11,28 +11,29 @@ process SAMTOOLS_VIEW {
...
     input:
     tuple val(meta), path(input)
     path fasta
 
     output:
-    tuple val(meta), path("*.bam") , optional: true, emit: bam
-    tuple val(meta), path("*.cram"), optional: true, emit: cram
-    path  "versions.yml"                           , emit: versions
+    tuple val(meta), path("*.bam") , emit: bam , optional: true
+    tuple val(meta), path("*.cram"), emit: cram, optional: true
```
Comment (maxulysse) [inline modules/samtools/view/main.nf:27]: so `optional: true` in the end, I like that

Comment (FriederikeHanssen) [reply modules/samtools/view/main.nf:27]: Thats what I saw harshil do in a bunch of PRs on monday.

File: modules/samtools/fastq/main.nf [hunk 2]
```diff
@@ -32,7 +32,7 @@ process SAMTOOLS_FASTQ {
     """
     samtools fastq \\
         $options.args \\
-        -@ $task.cpus \\
+        --threads ${task.cpus}-1 \\
```
Comment (phue) [inline modules/samtools/fastq/main.nf:35]:
```suggestion
        --threads ${task.cpus-1} \\
```

File: modules/samtools/fixmate/main.nf [hunk 3]
```diff
@@ -33,7 +33,7 @@ process SAMTOOLS_FIXMATE {
     samtools \\
         fixmate  \\
         $options.args \\
-        -@ $task.cpus \\
+        --threads ${task.cpus}-1 \\
```
Comment (phue) [inline modules/samtools/fixmate/main.nf:36]:
```suggestion
        --threads ${task.cpus-1} \\
```

File: modules/samtools/flagstat/main.nf [hunk 4]
```diff
@@ -27,7 +27,7 @@ process SAMTOOLS_FLAGSTAT {
 
     script:
     """
-    samtools flagstat $bam > ${bam}.flagstat
+    samtools flagstat --threads ${task.cpus}-1 $bam > ${bam}.flagstat
```
Comment (phue) [inline modules/samtools/flagstat/main.nf:30]:
```suggestion
    samtools flagstat --threads ${task.cpus-1} $bam > ${bam}.flagstat
```

File: modules/samtools/index/main.nf [hunk 5]
```diff
@@ -29,7 +29,8 @@ process SAMTOOLS_INDEX {
 
     script:
     """
-    samtools index $options.args $input
+    samtools index -@ ${task.cpus}-1 $options.args $input
```
Comment (phue) [inline modules/samtools/index/main.nf:32]:
```suggestion
    samtools index -@ ${task.cpus-1} $options.args $input
```

File: modules/samtools/merge/main.nf [hunk 6]
```diff
@@ -32,7 +32,8 @@ process SAMTOOLS_MERGE {
     def file_type = input_files[0].getExtension()
     def reference = fasta ? "--reference ${fasta}" : ""
     """
-    samtools merge ${reference} ${prefix}.${file_type} $input_files
+    samtools merge --threads ${task.cpus}-1 $options.args ${reference} ${prefix}.${file_type} $input_files
```
Comment (phue) [inline modules/samtools/merge/main.nf:35]:
```suggestion
    samtools merge --threads ${task.cpus-1} $options.args ${reference} ${prefix}.${file_type} $input_files
```

File: modules/samtools/stats/main.nf [hunk 7]
```diff
@@ -29,7 +29,8 @@ process SAMTOOLS_STATS {
     script:
     def reference = fasta ? "--reference ${fasta}" : ""
     """
-    samtools stats ${reference} ${input} > ${input}.stats
+    samtools stats --threads ${task.cpus}-1 ${reference} ${input} > ${input}.stats
```
Comment (phue) [inline modules/samtools/stats/main.nf:32]:
```suggestion
    samtools stats --threads ${task.cpus-1} ${reference} ${input} > ${input}.stats
```

File: modules/samtools/view/main.nf [hunk 8]
```diff
@@ -11,28 +11,29 @@ process SAMTOOLS_VIEW {
...
+    path  "versions.yml"           , emit: versions
 
     script:
     def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
     def reference = fasta ? "--reference ${fasta} -C" : ""
     def file_type = input.getExtension()
     """
-    samtools view ${reference} $options.args $input > ${prefix}.${file_type}
+    samtools view --threads ${task.cpus}-1 ${reference} $options.args $input > ${prefix}.${file_type}
```
Comment (phue) [inline modules/samtools/view/main.nf:35]:
```suggestion
    samtools view --threads ${task.cpus-1} ${reference} $options.args $input > ${prefix}.${file_type}
```

**Review by phue — CHANGES_REQUESTED** [review]
Just a small change, but I think an important one.
We need to make sure that `task.cpus -1` is calculated before nextflow substitutes the variable.
Otherwise bad things could happen when samtools parses the flag (or it will simply ignore it)

Other than that, looks good to me :+1:

**Comment by FriederikeHanssen** [comment]
right, thanks for catching that

**Review by phue — APPROVED** [review]

**Pull request merged**